servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java

```diff
@@ -83,15 +119,61 @@ protected HttpHeaders payloadComplete(final HttpHeaders trailers) {
                     ensureGrpcStatusReceived(trailers);
                     return trailers;
                 }
+
+                @Override
+                protected HttpHeaders payloadFailed(final Throwable cause, final HttpHeaders trailers)
+                        throws Throwable {
+
+                    // local cancel
+                    if (cause instanceof CancellationException) {
+                        // include the cause so that caller can determine who cancelled request
+                        throw new GrpcStatusException(new GrpcStatus(CANCELLED, cause), () -> null);
+                    }
+
+                    // local timeout
+                    if (cause instanceof TimeoutException) {
+                        // omit the cause unless logging has been configured at TRACE
+                        throw new GrpcStatusException(
+                                LOGGER.isTraceEnabled() ?
+                                    new GrpcStatus(DEADLINE_EXCEEDED, cause) :
+                                    DEADLINE_EXCEEDED.status(),
```

Comment:
Why in case of `DEADLINE_EXCEEDED` the cause is not useful? The exception message provides a good idea abot what was the configured timeout value.
If you still prefer a condition based on the logger level, consider using `DEBUG` level. Users rarely use `TRACE` for all packages, usually only for specific logger names.

Reply:
The cause message is useful but the stack trace is probably not. In the examples I have seen the stack is quite large and there was not much point in ever potentially transmitting it or even keeping the stack trace. I did consider adding the cause message as the status details but that doesn't seem to be consistently used. I will switch it to DEBUG.